File: C++/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(vector<vector<int>> maps)
{
	int answer = 0;

	queue<pair<int, int>> queue;

	int n = maps.size();
	int m = maps[0].size();

	//if (maps[n - 2][m -1] == 0 && maps[n - 1][m - 2] == 0) return -1;

	vector<vector<int>> dist(n, vector<int>(m, -1));
	int dx[4] = { -1, 0, 1, 0 };
	int dy[4] = { 0, 1, 0, -1 };

	dist[0][0] = 1;

	queue.push({ 0, 0 });

	while (!queue.empty())
	{
		int x = queue.front().first;
		int y = queue.front().second;

		queue.pop();

		for (int i = 0; i < 4; i++)
		{
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (nx >= 0 && nx < n && ny >= 0 && ny < m && maps[nx][ny] == 1 && dist[nx][ny] == -1)
			{
				dist[nx][ny] = dist[x][y] + 1;
				queue.push({ nx, ny });
			}
		}
	}

	answer = dist[n - 1][m - 1];

	return answer;
}

File: Python3/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
def solution(maps):
    answer = 0
    
    h = len(maps)
    w = len(maps[0])
    
    queue = []
    visited = [[-1] * w for _ in range(h)]
    
    dir_x = [0, 1, 0, -1]
    dir_y = [1, 0, -1, 0]
    
    queue.append((0, 0))
    visited[0][0] = 1
    while queue:
        pair = queue.pop(0)
        org_x, org_y = pair[0], pair[1]
        
        for i in range(4):
            new_x = org_x + dir_x[i]
            new_y = org_y + dir_y[i]
            
            if 0 <= new_x < h and 0 <= new_y < w and maps[new_x][new_y] == 1 and visited[new_x][new_y] == -1:
                visited[new_x][new_y] = visited[org_x][org_y] + 1
                queue.append((new_x, new_y))
    
    answer = visited[-1][-1]
    
    return answer